### PR TITLE
Better virtualization

### PR DESCRIPTION
**Pull request by pablodanswer**
## How Has This Been Tested?
[Describe the tests you ran to verify your changes]

## Accepted Risk
[Any know risks or failure modes to point out to reviewers]

## Related Issue(s)
[If applicable, link to the issue(s) this PR addresses]

## Checklist:
- [ ] All of the automated tests pass
- [ ] All PR comments are addressed and marked resolved
- [ ] If there are migrations, they have been rebased to latest main
- [ ] If there are new dependencies, they are added to the requirements
- [ ] If there are new environment variables, they are added to all of the deployment methods
- [ ] If there are new APIs that don't require auth, they are added to PUBLIC_ENDPOINT_SPECS
- [ ] Docker images build and basic functionalities work
- [ ] Author has done a final read through of the PR right before merge

### CODE REVIEW

File: web/src/app/chat/ChatPage.tsx [hunk 1]
```diff
@@ -1133,7 +1141,9 @@ export function ChatPage({
 
       await delay(50);
       while (!stack.isComplete || !stack.isEmpty()) {
-        await delay(0.5);
+        if (stack.isEmpty()) {
```
Comment (Weves) [inline web/src/app/chat/ChatPage.tsx:1144]: use a more descriptive variable name

**Review by Weves — COMMENTED** [review]
Seems to be a little bit of a problem with the rendering of single `:

Screenshot 2024-10-01 at 7 49 56 PM

File: web/src/app/chat/ChatPage.tsx (hunk 1 above)
Comment (Weves) [reply web/src/app/chat/ChatPage.tsx:1144]: ^ is sam's comment, no need to change

**Review by Weves — APPROVED** [review]

File: web/src/app/chat/message/CodeBlock.tsx [hunk 2]
```diff
@@ -26,132 +28,99 @@ export const CodeBlock = memo(function CodeBlock({
...
+        <div className="flex items-center space-x-2">
+          <FiCopy size={16} />
+          <span>Copy code</span>
+        </div>
+      )}
+    </div>
+  ));
+  CopyButton.displayName = "CopyButton";
+
+  const CodeContent = memo(() => {
```
Comment (mfunkie) [inline web/src/app/chat/message/CodeBlock.tsx:59]: I would recommend creating these components outside of the render path and passing props to them.  This isn't going to work how you think it is.  In between re-renders an entirely new component will be created.

Comment (pablonyx) [reply web/src/app/chat/message/CodeBlock.tsx:59]: Very true and appreciate the callout! Actually meant to remove the `memo` from these internal components (like you mention, not actually memoizing right now) - the `MemoizedCodeLine` is really what boosted performance.

**Pull request merged**